storage/grpc_reader_multi_range.go

```diff
@@ -38,3 +51,834 @@ func (g *readIDGenerator) Next() int64 {
+		m.handleStreamEnd(result)
+		return
+	}
+
+	resp := result.resp
+	if handle := resp.GetReadHandle().GetHandle(); len(handle) > 0 {
+		m.lastReadHandle = handle
+		if m.params.handle != nil {
+			*m.params.handle = handle
+		}
+	}
+
+	m.attrsOnce.Do(func() {
+		if meta := resp.GetMetadata(); meta != nil {
+			obj := newObjectFromProto(meta)
+			attrs := readerAttrsFromObject(obj)
+			m.attrs = attrs
+			close(m.attrsReady)
+
+			for _, req := range m.pendingRanges {
+				if req.offset < 0 {
+					m.convertToPositiveOffset(req)
+				}
+			}
+		} else {
+			m.handleStreamEnd(mrdSessionResult{err: errors.New("storage: first response from BidiReadObject stream missing metadata")})
+		}
+	})
+
+	for _, dataRange := range resp.GetObjectDataRanges() {
+		readID := dataRange.GetReadRange().GetReadId()
+		req, exists := m.pendingRanges[readID]
+		if !exists || req.completed {
+			continue
+		}
+
+		content := dataRange.GetChecksummedData().GetContent()
+		req.bytesWritten += int64(len(content))
+		_, err := req.output.Write(content)
+		if err != nil {
+			m.failRange(req, err)
+			continue
+		}
+
+		if dataRange.GetRangeEnd() {
+			req.completed = true
+			delete(m.pendingRanges, req.readID)
+			m.runCallback(req.origOffset, req.bytesWritten, nil, req.callback)
+		}
+	}
+}
+
+// ensureSession is now only for reconnecting *after* the initial session is up.
+func (m *multiRangeDownloaderManager) ensureSession(ctx context.Context) error {
+	if m.currentSession != nil {
+		return nil
+	}
+	if m.permanentErr != nil {
+		return m.permanentErr
+	}
+
+	// Using run for retries
+	return run(ctx, func(ctx context.Context) error {
+		if m.currentSession != nil {
+			return nil
+		}
+		if m.permanentErr != nil {
+			return m.permanentErr
+		}
+
+		session, err := newBidiReadStreamSession(m.ctx, m.sessionRespC, m.client, m.settings, m.params, proto.Clone(m.readSpec).(*storagepb.BidiReadObjectSpec))
+		if err != nil {
+			redirectErr, isRedirect := isRedirectError(err)
+			if isRedirect {
+				m.readSpec.RoutingToken = redirectErr.RoutingToken
+				m.readSpec.ReadHandle = redirectErr.ReadHandle
+				return fmt.Errorf("%w: %v", errBidiReadRedirect, err)
+			}
+			return err
+		}
+		m.currentSession = session
+
+		var rangesToResend []*storagepb.ReadRange
+		for _, req := range m.pendingRanges {
+			if !req.completed {
+				readLength := req.length
+				if req.length > 0 {
+					readLength -= req.bytesWritten
+				}
+				if readLength < 0 {
+					readLength = 0
+				}
+
+				if req.length == 0 || readLength > 0 {
+					rangesToResend = append(rangesToResend, &storagepb.ReadRange{
+						ReadOffset: req.offset + req.bytesWritten,
+						ReadLength: readLength,
+						ReadId:     req.readID,
+					})
+				}
+			}
+		}
+		if len(rangesToResend) > 0 {
+			m.currentSession.SendRequest(&storagepb.BidiReadObjectRequest{ReadRanges: rangesToResend})
+		}
+		return nil
+	}, m.settings.retry, true)
+}
+
+var errBidiReadRedirect = errors.New("bidi read object redirected")
```

Comment:
Should this error have more or different information when surfaced to the user? This doesn't seem actionable